### ISSUE TITLE
Bayesian updating of hyperparameters

### ISSUE THREAD

**Issue by milicasan** [SheffieldML/GPyOpt#44]
Dear Developers,
We were able to customise our kernel and BO.model, and used it with run_optimisation. The BO procedure works well, but we noticed that if we update our model with each acquisition, this also updates the hyperparameters of the GP process (maximum likelihood) and often results in overfitting. Increasing model update interval was not a good idea either. We cannot find how to a) impose upper/lower numerical bounds on the hyperparameter search and b) refit the model with each acquisitions but update the model hyperparameters more slowly. Ideally, we would use Bayesian updating of hyperparameters (with priors on the hyperparameters), is this available? This is currently the biggest issue for us.
Many thanks for your assistance, Milica.

**Comment by 1Reinier** [comment]
Have you tried MCMC for the GP update? That should take care of it.

**Comment by milicasan** [comment]
Hi Reinier, 
thanks for the suggestion. Yes, it looks like this may be the only option, but were hoping for other choices. We've been working with the gpml package in the past year and we never needed to go to MCMC... If you're a developer, can you tell us what else other than basic and MCMC is available at present, even experimentally? 
Best, Milica

**Comment by 1Reinier** [comment]
Sorry, I am not a developer for this package

**Comment by javiergonzalezh** [comment]
Hi Milica,

You can do MCMC on the hyper-parameters of the GP model and integrate the acquisitions with respect to those.  In this notebook you have the details of how to do it. 

https://github.com/SheffieldML/GPyOpt/blob/master/manual/GPyOpt_integrating_model_hyperparameters.ipynb

Here you have some guidance of how to deal with the parameters in the GPy models

http://nbviewer.jupyter.org/github/SheffieldML/notebook/blob/master/GPy/sampling_hmc.ipynb

Hope it helps.
Javier

**Comment by milicasan** [comment]
Hi Javier, 
Thanks for this info, we're already working on the MCMC and your links will be really useful. Do you have any plans to add functionality here, or are your priorities elsewhere? We would also explore the possibility of modifying the basic model update (max likelihood) to introduce an interval for the update of the hyperparameters, this simple step has really helped us in the past. Would this be of any interest to you?
Best, Milica.

**Comment by javiergonzalezh** [comment]
In GPy you can constraint the hyper-parameters to some given interval already.  Have a look to this notebook to see how to interact with the hyper-parameters of the models:

http://nbviewer.jupyter.org/github/SheffieldML/notebook/blob/master/GPy/models_basic.ipynb
You can force them to be positive, sit on certain interval, etc.

**Comment by milicasan** [comment]
Hi Javier, 
thanks for your help on this topic, it really helped us to understand the code better. We now learned to separate model update from model prediction: now we can predict with each acquisition and update with an interval, and this has stabilised our fitting considerably (if you wish, we can contribute this module). We also tried the MCMC, but gained no further benefit: we'll reconsider this if the things get really hairy in the future.
All the best, Milica.
